text/3305-static-in-pattern.md

```diff
@@ -0,0 +1,266 @@
+
+# Motivation
+
+[motivation]: #motivation
+
+Rust pattern matches compare a scrutinee against compile-time information. Rust generally doesn't allow patterns to depend on runtime information; that is relegated to match guards. However, Rust currently also prevents patterns from depending on *link-time* information, specifically statics from `extern` blocks.
+
+I encountered this restriction while trying to port the Rust standard library to [cosmopolitan libc](https://justine.lol/cosmopolitan/index.html). Cosmopolitan provides an API that mostly matches POSIX, with one major exception: constants like `ENOSYS` and `EINVAL`, which on most platforms are defined as C `#define`s (equivalent to Rust `const`s), are instead provided as C `const`s (equivalent to Rust non-`mut` `static`s).
+
+```rust
+// libc crate
+
+cfg_if! {
+    if #[cfg(target_env = "cosmopolitan")] {
+        extern "C" {
+            pub static EINVAL: i32;
+            pub static ENOSYS: i32;
+            pub static ENOENT: i32;
+        }
+    } else {
+        pub const EINVAL: i32 = 42;
+        pub const ENOSYS: i32 = 43;
+        pub const ENOENT: i32 = 44;
+    }
+}
+
+// stdlib code
+
+use libc::*;
+
+fn process_error(error_code: i32) {
+    match error_code {
+        // Compiler throws error EO530 on Cosmopolitan,
+        // because `static`s can't be used in patterns, only `const`s
+        EINVAL => do_stuff(),
+        ENOSYS => panic!("oh noes"),
+        ENOENT => make_it_work(),
+        _ => do_different_stuff(),
+    }
+}
+```
+
+Because Rust patterns don't support statics, all the `match` expressions in the standard library that refer to POSIX constants would currently need to be rewritten to accommodate Cosmopolitan.
+
+```rust
+// stdlib code adapted for cosmopolitan
+
+use libc::*;
+
+fn process_error(error_code: i32) {
+    if error_code == EINVAL {
+        do_stuff();
+    } else if error_code == ENOSYS {
+        panic!("oh noes");
+    } else if error_code == ENOENT {
+        make_it_work();
+    } else {
+        do_different_stuff();
+    }
+}
+```
+
+Needless to say, this is unlikely to ever be upstreamed. Allowing statics in patterns would solve this use-case much more cleanly.
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+Rust patterns can refer to constants.
+
+```rust
+const EVERYTHING: i32 = 42;
+
+fn foo(scrutinee: i32) {
+    match scrutinee {
+        EVERYTHING => println!("have all of it"),
+        _ => println!("need moar"),
+    }
+}
+```
+
+With this feature, they can refer to statics as well.
+
+```rust
+static EVERYTHING: i32 = 42;
+
+fn foo(scrutinee: i32) {
+    match scrutinee {
+        EVERYTHING => println!("have all of it"),
+        _ => println!("need moar"),
+    }
+}
+```
+
+Mutable statics are not allowed, however. Patterns can't reference information that can change at runtime, and also can't be `unsafe`.
```

Comment:
Ignoring the `unsafe` issue, is there a reason that the pattern can't reference information that can change at runtime? My understanding of how cosmopolitan is going to use this _is_ that it references information that changes at runtime (but only once, before `main`), the values of the statics are not available at link time like they normally are. Given that why would it not be possible for a developer to do the same thing using `static mut` where they set the value once early in `main` before the pattern is ever tested against.

Reply:
> Is there a reason that the pattern can't reference information that can change at runtime? 

The "patterns can't reference information that can change at runtime" rule is not something that I am proposing, it's a rule that Rust already has and that I see no reason to change. Tearing down this Chesterton's Fence may or may not be a good idea, but is out of scope for this RFC.

> My understanding of how cosmopolitan is going to use this is that it references information that changes at runtime (but only once, before main)

This is where the exact definition of "at runtime" comes into play. If you define "runtime" as "the time when Rust code runs", then the values of the statics don't change at runtime. I argue that for the purposes of this feature, it doesn't really matter whether the values of the statics are set by the linker, the operating system, the C runtime, etc; all of these things fall into the common category of "`rustc` doesn't know what the value is, but Rust code can rely on the value being constant." Note that Cosmopolitan sets the values of the statics in assembly (that I don't presume to fully understand); C code written against Cosmopolitan sees them as standard C `extern const`s, and doesn't need to worry about their value changing, either.

Reply:
Fundamentally this is how all rust static values work. Officially Rust "doesn't have life before `main`", but it still requires *someone* (an OS or an assembly boot script) to initialize all the pre-conditions before Rust code actually begins executing.

So this part of the RFC is basically nothing new.

Reply:
I feel a good analogy is dynamic linking. When you call a dynamically linked function in Rust, the compiler doesn't know the address of that function, it only gets determined when the program runs. But you can still write a safe wrapper for the dynamically linked function, and then treat it pretty much like any other Rust function (though not like a `const fn`, of course).

Extern statics are the same. Unlike `const`s, `rustc` doesn't know their value. But, once the necessary `#[trusted_static]` ceremonies are performed, it should still be possible (IMO) to treat them"pretty much" like any other constant value.

Reply:
> Fundamentally this is how all rust static values work.

Not normally, the vast majority of Rust code is statically linked which means the values are known at link time (and LTO would then be able to make `const` and `static` patterns generate the same code).



> Extern statics are the same. Unlike `const`s, `rustc` doesn't know their value. But, once the necessary `#[trusted_static]` ceremonies are performed, it should still be possible (IMO) to treat them"pretty much" like any other constant value.

Would it be unsound to use `#[trusted_static]` on a dynamically linked extern static that gets initialized early-in-`main` before any code that reads the static runs? It just seems like it should be possible to unsafely promise that a value is constant once it starts being used in patterns, given that is essentially the same behavior you get with dynamically linked statics (where the dynamic loader has unsafely promised to the executable that all statics have been initialized and will no longer be written to).

Reply:
> Not normally, the vast majority of Rust code is statically linked which means the values are known at link time (and LTO would then be able to make const and static patterns generate the same code).

Okay, so yeah there's a missed optimization compared to a different situation with looser constraints. That's the same as how you can't inline an external function. A mild shame, but just an unavoidable performance difference. It doesn't impact correctness.

Reply:
Oh whoops, *I'm* the one in charge of trusted statics? That kinda stalled out because of like, long term correctness discussion XD

uhm, don't expect that one soon.

Reply:
> That kinda stalled out because of like, long term correctness discussion

Is there a list of the unresolved questions somewhere? (Or does one need to be compiled?) Might it be possible to ship an obviously correct MVP sooner, and consider loosening safety requirements later?

Reply:
Just the zulip thread, which the project tracking issue should link to.

The problem is really simple: It's hard to tell people that the unsafe-adjacent code they've been writing the past 6 years is just barely wrong, and that they need to start doing everything the opposite way because of small edge cases. That's a hard sell.

Reply:
> It's hard to tell people that the unsafe-adjacent code they've been writing the past 6 years is just barely wrong, and that they need to start doing everything the opposite way because of small edge cases.

Could we figure out the shiny new Right Way of Doing Things ® now, worry about migrating/accommodating old code later once people are used to the new syntax? The longer we wait before offering better tools, the more technically-bad code gets written, and the more painful any necessary changes will be.

Reply:
Yeah, it's easy if migration of code and re-training of programmers isn't a concern ;3

Literally we just would have to say one day "hey declaring extern static/fn is unsafe now". And like, a compiler patch to enforce that. That's the entire end goal for Rust 3015. It's just the migration that needs to be designed really.